thrust/thrust/system/detail/generic/logical.h

```diff
@@ -42,14 +40,16 @@ template <typename ExecutionPolicy, typename InputIterator, typename Predicate>
 _CCCL_HOST_DEVICE bool
 all_of(thrust::execution_policy<ExecutionPolicy>& exec, InputIterator first, InputIterator last, Predicate pred)
 {
-  return thrust::find_if(exec, first, last, thrust::detail::not1(pred)) == last;
+  // TODO(bgruber): we could implement this even better using an early exit
+  return thrust::count_if(exec, first, last, thrust::detail::not1(pred)) == 0;
 }
 
 template <typename ExecutionPolicy, typename InputIterator, typename Predicate>
 _CCCL_HOST_DEVICE bool
 any_of(thrust::execution_policy<ExecutionPolicy>& exec, InputIterator first, InputIterator last, Predicate pred)
 {
-  return thrust::find_if(exec, first, last, pred) != last;
+  // TODO(bgruber): we could implement this even better using an early exit
+  return thrust::count_if(exec, first, last, pred) > 0;
```

Comment:
important: using reduction has performance advantages, but also introduces significant performance regressions compared to short circuiting that we have today. If there's a match (or mismatch, depending on the algorithm) early in the sequence, there's no need to read the remaining vector or perform comparison operator. Here are some performance results:

<img width="652" alt="image" src="https://github.com/NVIDIA/cccl/assets/9890394/ef777016-a85d-4009-9992-ce9543abc208">

Performance data above compares find vs count on 2^28 int32s. Since int32 is easy to load and comparison is fast to perform, we see that count becomes faster if the match (mismatch) is after 4% of the input size. But if we compare different vectors, find will bump into difference immediately and we won't load the reminder of the vector. This makes find ~9x faster in this scenario. 

But both speedup and inflection point depend on the price of loading data + price of computing the binary operator. Here is the same comparison for double with a bit more fancy comparison operator (comparing absolute difference with an epsilon):

<img width="672" alt="image" src="https://github.com/NVIDIA/cccl/assets/9890394/832b328a-2440-44e0-a2b1-bf180e95f90f">

Here we already see 16x slowdown from transitioning to count, and the inflection point shifted much further to 6% of the input. 

The performance regressions seem to be severe enough for us not to rush with merging this PR. We wanted to ask @gonidelis to explore a bit different reduction algorithm, where we'd store atomic flag upon satisfying a condition, so that later thread blocks would early exist instead of loading data. The idea is that this approach would get us more moderate performance regressions in mentioned cases while providing comparable speedups.

Reply:
Btw, I took these two issues, #720 and #712, as inspiration to go for `count_if`, since both of them claim that it showed a several-factor improvement. It also made a lot of sense to me to route the implementation through `all_of`, since `all_of` does not need to compute any kind of "first occurence" of a predicate match, like `find_if`, and `all_of` would not require reading the entire input, like `count_if`. So if you really want to block this PR, I would like to at least merge the reimplementation of `equals` based on `all_of` and not `mismatch` (which implies an order).